src/cdk/overlay/keyboard/overlay-keyboard-dispatcher.ts

```diff
@@ -0,0 +1,98 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Injectable, Optional, SkipSelf, OnDestroy} from '@angular/core';
+import {OverlayRef} from '../overlay-ref';
+import {Observable} from 'rxjs/Observable';
+import {Subject} from 'rxjs/Subject';
+import {RxChain, filter, takeUntil} from '@angular/cdk/rxjs';
+import {fromEvent} from 'rxjs/observable/fromEvent';
+import {merge} from 'rxjs/observable/merge';
+
+/**
+ * Service for dispatching keyboard events that land on the body to appropriate overlay ref,
+ * if any. It maintains a list of attached overlays to determine best suited overlay based
+ * on event target and order of overlay opens.
+ */
+@Injectable()
+export class OverlayKeyboardDispatcher implements OnDestroy {
+
+  /** Currently attached overlays in the order they were attached. */
+  _attachedOverlays: OverlayRef[] = [];
+
+  /** Emits when the service is destroyed. */
+  private _onDestroy = new Subject<void>();
+
+  constructor() {
+    this._dispatchKeyboardEvents();
+  }
+
+  ngOnDestroy() {
+    this._onDestroy.next();
+    this._onDestroy.complete();
+  }
+
+  /** Add a new overlay to the list of attached overlay refs. */
+  add(overlayRef: OverlayRef): void {
+    this._attachedOverlays.push(overlayRef);
+  }
+
+  /** Remove an overlay from the list of attached overlay refs. */
+  remove(overlayRef: OverlayRef): void {
+    const index = this._attachedOverlays.indexOf(overlayRef);
+    if (index > -1) {
+      this._attachedOverlays.splice(index, 1);
+    }
+  }
+
+  /**
+   * Subscribe to all key events that land on the body and dispatch those
+   * events to the appropriate overlay.
+   */
+  private _dispatchKeyboardEvents(): void {
+    const bodyKeyboardEvents = merge(
+        fromEvent(document.body, 'keydown'),
+        fromEvent(document.body, 'keypress'),
```

Comment:
I'm not sure that this should handle `keydown`, `keypress` and `keyup` @jelbourn. Since for 99% of cases you'd only need `keydown`, it means that every time you consume the service, you'd have to have a `filter(event => event.type === 'keydown')`.

Reply:
I'd be happy with this. I was aiming for generic, but the api could be "rebranded" 'to just `keydownEvents`.